Mathlib/Analysis/InnerProductSpace/Spectrum.lean

```diff
@@ -108,6 +108,31 @@ theorem orthogonalComplement_iSup_eigenspaces (hT : T.IsSymmetric) (μ : 𝕜) :
   have H₂ : eigenspace T μ ⟂ p := (Submodule.isOrtho_orthogonal_right _).mono_left (le_iSup _ _)
   exact H₂.disjoint
 
+/-- Every generalized eigenspace of a symmetric operator is an eigenspace. -/
+theorem genEigenspace_eq_eigenspace
+    {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric) {n : ℕ} {μ : 𝕜} (hn : 1 ≤ n) :
+    genEigenspace T μ n = genEigenspace T μ 1 := by
+  refine Nat.le_induction rfl (fun k hk ih ↦ ?_) n hn
+  refine ih ▸ le_antisymm (fun x hx ↦ ?_) ((genEigenspace T μ).mono k.le_succ)
+  obtain (rfl | hx_ne) := eq_or_ne x 0
+  · exact zero_mem _
+  · have hμ : starRingEnd 𝕜 μ = μ := hT.conj_eigenvalue_eq_self <|
+    hasEigenvalue_of_hasGenEigenvalue (k := k + 1) <|
+    (genEigenspace T μ (k + 1)).ne_bot_iff.mpr ⟨x, hx, hx_ne⟩
+    have hTμ : ((T - μ • 1) ^ k).IsSymmetric := by
+      aesop (add simp LinearMap.one_eq_id)
+    rw [mem_genEigenspace, ← norm_eq_zero, ← sq_eq_zero_iff, norm_sq_eq_inner (𝕜 := 𝕜),
+      hTμ, ← LinearMap.comp_apply, ← LinearMap.mul_eq_comp, ← pow_add]
+    simp [mem_genEigenspace .. |>.mp <| (genEigenspace T μ).mono (show k + 1 ≤ k + k by gcongr) hx]
+
+lemma maxGenEigenspace_eq_eigenspace {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric)
+    {μ : 𝕜} : maxGenEigenspace T μ = eigenspace T μ := calc
+  _ = ⨆ n, genEigenspace T μ (n + 1) := by
+    rw [maxGenEigenspace_def, ← sup_iSup_nat_succ, genEigenspace_def]; simp [LinearMap.one_eq_id]
+  _ = ⨆ _ : ℕ, genEigenspace T μ 1 := by
+    congr! 2 with n; exact genEigenspace_eq_eigenspace hT n.succ_pos
+  _ = eigenspace T μ := by simp [genEigenspace_def, eigenspace_def]
+
```

Comment:
Bearing in mind the [#Module.End.IsSemisimple.genEigenspace_eq_eigenspace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Semisimple.html#Module.End.IsSemisimple.genEigenspace_eq_eigenspace) and [Module.End.IsSemisimple.maxGenEigenspace_eq_eigenspace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Semisimple.html#Module.End.IsSemisimple.maxGenEigenspace_eq_eigenspace) I wonder if instead of these we could prove:
```lean
theorem isSemisimple {T : Module.End 𝕜 E} (hT : T.IsSymmetric) :
    T.IsSemisimple := by
  refine isSemisimple_iff.mpr fun p hp ↦ ⟨pᗮ, fun x hx ↦ ?_, IsCompl.mk ?_ ?_⟩
  · simp only [Submodule.mem_comap, Submodule.mem_orthogonal] at hx ⊢
    intro y hy
    rw [← hT y x]
    exact hx (T y) (hp hy)
  · rw [disjoint_iff]
    exact Submodule.inf_orthogonal_eq_bot p
  · rw [codisjoint_iff]
    have : FiniteDimensional 𝕜 E := sorry -- Can avoid I presume
    apply Submodule.sup_orthogonal_of_completeSpace -- Overkill I think
```

Reply:
OK. Will work on this a bit.

Reply:
In infinite dimensions, you do seem to need that a subspace `K` is closed in order for `K` perp and `K` to span the space...so I don't think what you have there is overkill...

Reply:
@ocfnash I'm not sure why completeness is overkill here (although finite dimensionality is, but we run into the problem that [FiniteDimensional.complete](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Module/FiniteDimension.html#FiniteDimensional.complete) is not an instance, which is slightly annoying whenever it comes up). And `T := LinearMap.id` is symmetric. Moreover, it is semisimple if and only if every subspace (not necessarily *closed*!) is complemented. I think this isn't true for inner product spaces that are not complete.

Reply:
> It isn't true if the space isn't complete.

In that case, just assume completeness. When users call this lemma, they can invoke the [FiniteDimensional.complete](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Module/FiniteDimension.html#FiniteDimensional.complete) lemma if necessary.

Reply:
I should have been more precise. To get the semisimplicity here we need that all subspaces are complete, since we want them all to be orthogonally complemented. This requires the space to be finite dimensional...or otherwise there are going to be non-closed subspaces around...

Reply:
For this reason, though, we probably want to keep those other results, as well...because this one only holds in finite dimensions...

Reply:
Yes, I was in a hurry earlier (about to start work): of course you need some assumption because in general there will be non-closed invariant subspaces, or else you need to adjust the conclusion. One idea I had in mind as I was writing this were that one could consider introducing a "semisimple on finite-dimensional subspaces" definition. This is weaker than `IsSemisimple` and is satisfied here and I think should still be enough for `IsSemisimple.maxGenEigenspace_eq_eigenspace` to hold. I'm not sure if I really like this idea though.

Annoyingly I'm still under time pressure so I can't think properly about this till later this evening.

Reply:
The other thing I did earlier was to put together https://github.com/leanprover-community/mathlib4/pull/17654

This is the other half of item 3 in [my comment here](https://github.com/leanprover-community/mathlib4/pull/16569#issuecomment-2372447382) where I said "(and a similar statement for independence)".

If this is merged, then we will know that the simultaneous maximum generalised eigenspaces of any family of endomorphims in finite dimensions (over any field) decompose the space as a direct sum.

Pursuing the approach not to specialise the coefficients prematurely, the next thing to do would be to prove the independent result that if the members of the family are self-adjoint (in the sense of [LinearMap.IsSelfAdjoint](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/SesquilinearForm.html#LinearMap.IsSelfAdjoint)) wrt a non-degenerate bilinear form then generalised eigenspaces are eigenspaces.

And then that's it really (all just algebra) except a very short lemma in the analysis corner of the library which restates the final conclusion over an inner-product space.

Reply:
> the simultaneous maximum generalised eigenspaces of any family of endomorphims in finite dimensions (over any field) decompose the space as a direct sum.

I think this requires that the characteristic polynomials all split (or else that the field is algebraically closed). Indeed, generalized eigenvalues are eigenvalues, so this is necessary.

Reply:
@ocfnash : This is now touching a lot of files. Would it be ok to pursue these latest ideas in another PR?

Reply:
I'll let Jireh make the call here. I don't think the number of files is large (currently it's 5, and could easily be 3).

The only point about which I feel somewhat strongly is that we should not be proving results with the assumptions of `ℝ` or `ℂ` when they are true for any field.

However, again, I'll leave this in Jireh's hands.

Reply:
Let me talk to Jireh about it @ocfnash. I guess you guys intend to add an assumption about the splitting of the charpoly, and that would do it...let's see what he says.